lib/logflare/backends.ex

```diff
@@ -478,21 +479,62 @@ defmodule Logflare.Backends do
     end
   end
 
+  def push_recent_events(%Source{}, []), do: :ok
+
+  def push_recent_events(%Source{id: source_id}, log_events) do
+    existing = :ets.lookup(:recent_events, source_id)
+
+    :ets.delete(:recent_events, source_id)
+
+    # sort newest to oldest
+    objects =
+      for event <- Enum.sort_by(log_events, fn %{body: %{"timestamp" => ts}} -> ts end, &<=/2) do
+        # don't cache source information
+        {source_id, %{event | source: nil}}
+      end
+
+    :ets.insert(:recent_events, Enum.take(objects ++ existing, 100))
```

Comment:
Can maybe sort and take 100 in one go with reduce?

Reply:
The existing items don't need to be sorted as it is ordered_set, sorting only the incoming reduces need to double sort

Reply:
Doing it separately cuts down reductions when there is a full buffer